design-proposals/klip-3-serialization-of-single-fields.md

```diff
@@ -0,0 +1,1076 @@
+-- the serialized keys and values will be unwrapped.
+CREATE SOURCE SK_SOURCE AS SELECT ID KEY, NAME STRING FROM S;
+ ```
+ 
+### INSERT
+
+Insert statements do not create sources of their own. 
+They use the serialization settings of their sink.
+
+For example,
+
+```sql
+-- will use the serialization settings of 'SINK` to determine if value should be wrapped or not.
+INSERT INTO SINK SELECT ID FROM SOURCE;
+
+-- will also use the serialization settings of `SINK`, though in this case they will be wrapped as its multi-field.
+INSERT INTO SINK SELECT ID, NAME FROM SOURCE;
+ ```
+ 
+### INSERT VALUES
+
+Insert values statements do not create sources of their own.
+They use the serialization settings of their sink.
+ 
+```sql
+-- will use the serialization settings of 'SINK` to determine if single field schemas should be wrapped or not.
+INSERT INTO SINK (ID) VALUES (10);
+
+-- will also use the serialization settings of `SINK`, though in this case they will be wrapped as its multi-field.
+INSERT INTO SINK VALUES (10, 'bob');
+ ```
+ 
+
+## Test plan
+
+Tests will be added to cover all valid combinations and permutations of 
+the following dimensions:
+
+* Format: `JSON` and `AVRO`
+* `ksql.persistence.wrap.single.keys` config?: `true` and `false` 
+* `ksql.persistence.wrap.single.values` config?: `true` and `false`
+* Source type: `TABLE` and `STREAM`
+* Source key: unwrapped single field, wrapped single field, multiple fields
+* Source value: unwrapped single field, wrapped single field, multiple fields
+* Source `WRAP_SINGLE_KEY`: `true` and `false` 
+* Source `WRAP_SINGLE_VALUE`: `true` and `false` 
+* Source `WRAP_SINGLE_FIELDS`: `true` and `false` 
+* Query type: `TABLE` and `STREAM`
+* Query `WRAP_SINGLE_KEY`: `true` and `false` 
+* Query `WRAP_SINGLE_VALUE`: `true` and `false` 
+* Query `WRAP_SINGLE_FIELDS`: `true` and `false`
+* Query key schema: single and multiple fields
+* Query value schema: single and multiple fields
+
+This will initially be done only for values. 
+Key support will be added as part of the structured keys work.
+
+Note: combining `WRAP_SINGLE_FIELDS` with either `WRAP_SINGLE_KEY` or
+`WRAP_SINGLE_VALUE` should result in a duplicate setting error.
+
+JSON tests will also be added to ensure the new configurations can be
+set via the `SET` command and are picked as expected.
+
+Existing tests are sufficient to ensure these changes do not effect the 
+format or schema of internal topics.
+
+## Documentation Updates
+
+* The server configuration documentation in `config-reference.rst` will 
+have the following two settings added:
+
+```rst
+.. _ksql-persistence-wrap-single-keys:
+
+---------------------------------
+ksql.persistence.wrap.single.keys
+---------------------------------
+
+Sets the default value for the ``WRAP_SINGLE_KEY`` property if one is
+not supplied explicitly in :ref:`CREATE TABLE <create-table>`,
+:ref:`CREATE STREAM <create-stream>`, :ref:`CREATE TABLE <create-table-as-select>`
```

Comment:
```suggestion
:ref:`CREATE STREAM <create-stream>`, :ref:`CREATE TABLE AS SELECT <create-table-as-select>`
```
Also three more instances of this typo below.

Reply:
well spotted :D